odl/datasets/util.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Utilities for ODL datasets."""
+
+import os
+from os.path import join, expanduser, exists
+
+try:
+    # Python 2
+    from urllib2 import urlopen
+except ImportError:
+    # Python 3+
+    from urllib.request import urlopen
+    
+from shutil import copyfileobj
+from scipy import io
+
+
+__all__ = ('get_data_dir', 'get_data')
+
+
+def get_data_dir():
+    """Get the data directory."""
+    data_home = expanduser(join('~', 'odl_data'))
+    if not exists(data_home):
+        os.makedirs(data_home)
+    return data_home
+
+
+def get_data(filename, subset, url):
+    """Get a dataset with from a url with local caching.
+    
+    Parameters
+    ----------
+    filename : str
+        Name of the file, for caching.
+    subset : str
+        To what subset the file belongs (e.g. 'ray_transform'). Each subset
+        is saved in a separate subfolder.
+    url : str
+        url to the dataset online.
+        
+    Returns
+    -------
+    dataset : dict
+        Dictionary containing the dataset.
+    """
+
+    # check if this data set has been already downloaded
+    data_dir = get_data_dir()
+    data_dir = join(data_dir, subset)
+    if not exists(data_dir):
+        os.makedirs(data_dir)
+
+    filename = join(data_dir, filename)
+
+    # if the file does not exist, download it
+    if not exists(filename):
+        print('data {}/{} missing, downloading from {}'
+              ''.format(subset, filename, url))
+        
+        # open the url of the data
+        with urlopen(url) as data_url:
```

Comment:
I meant [this comment](https://github.com/odlgroup/odl/pull/992#discussion_r115684409) but couldn't reference it during review.